UI/window-basic-main.cpp

```diff
@@ -167,6 +167,117 @@ static int CountVideoSources()
 	return count;
 }
 
+static void midicallback(double deltatime, std::vector<uint8_t> *message,
+		void *userData)
+{
+	UNUSED_PARAMETER(userData);
+	QMidiEvent *midiEvent = nullptr;
+	size_t nBytes = message->size();
+
+	if (nBytes > 0) {
+		midiEvent = new QMidiEvent(*message, deltatime);
+		blog(LOG_DEBUG, "midi message (%s):",
+				obs_key_to_name(midiEvent->getKey()));
+	} else {
+		return;
+	}
+
+	for (size_t i = 0; i < nBytes; i++) {
+		uint8_t val = message->at(i);
+		blog(LOG_DEBUG, "%i (0x%x)", val, val);
+	}
+
+	blog(LOG_DEBUG, "midi timestamp: %f\n", deltatime);
+
+	/* todo: figure out how to post events properly */
+	QWidget *widget = App()->focusWidget();
+	if (widget) {
+		const char* n = widget->metaObject()->className();
+		if (strcmp(n, "OBSHotkeyEdit") == 0)
+			App()->postEvent(widget, midiEvent);
+		else
+			App()->postEvent(App()->activeWindow(), midiEvent);
+	} else {
+		widget = App()->activeWindow();
+		if (widget)
+			App()->postEvent(widget, midiEvent);
+		else
+			App()->postEvent(App()->GetMainWindow(), midiEvent);
+	}
+}
+
+int obs_get_midi_ports()
```

Comment:
Style: C++ function names should be CamelCase.  C functions use snake_case.

Reply:
OBSGetMidiPorts or ? How should that go, b/c there's two acronyms in that function.

Reply:
Why does it need to be prefixed with "OBS"?  I've generally only seen that in class names.  GetMidiPorts or GetMIDIPorts should be fine.  Sometimes acronyms in CamelCase names get separated by an underscore, but check the project source for examples, or wait for Jim to render an opinion.

Reply:
To further elaborate on function names prefixed with "obs_", that's usually done in libobs (and maybe the frontend-api), not the UI.

Honestly, in function and variable names, "Midi" (or "midi" when camelCase instead of CamelCase) should be fine.